telemetry/statsd/src/main/scala/io/buoyant/telemetry/statsd/StatsDStatsReceiver.scala

```diff
@@ -24,25 +24,40 @@ private[telemetry] class StatsDStatsReceiver(
   val repr: AnyRef = this
 
   private[statsd] def flush(): Unit = {
-    gauges.values.foreach(_.send)
+    gauges.values.asScala.foreach(_.send)
   }
   private[statsd] def close(): Unit = statsDClient.stop()
 
-  private[this] val gauges = new ConcurrentHashMap[String, Metric.Gauge].asScala
+  private[this] val counters = new ConcurrentHashMap[String, Metric.Counter]
+  private[this] val gauges = new ConcurrentHashMap[String, Metric.Gauge]
+  private[this] val stats = new ConcurrentHashMap[String, Metric.Stat]
 
   protected[this] def registerGauge(name: Seq[String], f: => Float): Unit = {
+    deregisterGauge(name)
+
     val statsDName = mkName(name)
-    gauges(statsDName) = new Metric.Gauge(statsDClient, statsDName, f)
+    val _ = gauges.put(statsDName, new Metric.Gauge(statsDClient, statsDName, f))
   }
 
   protected[this] def deregisterGauge(name: Seq[String]): Unit = {
     val _ = gauges.remove(mkName(name))
   }
 
-  def counter(name: String*): Counter =
-    new Metric.Counter(statsDClient, mkName(name), sampleRate)
-
-  def stat(name: String*): Stat =
-    new Metric.Stat(statsDClient, mkName(name), sampleRate)
+  def counter(name: String*): Counter = {
+    val statsDName = mkName(name)
+    counters.putIfAbsent(
+      statsDName,
+      new Metric.Counter(statsDClient, statsDName, sampleRate)
+    )
+    counters.get(statsDName)
```

Comment:
well, we're already doing the potentially unnecessary allocation anyway: putIfAbsent takes arguments by value, not by name.  To avoid this we could switch to computeIfAbsent but that's a bit more of a pain.  Unclear if the optimization is worth it.

The thing that I care about here is atomicity.  As written, the value in the hashmap could change after putIfAbsent is called but before get is called.